lib/medusa/adapter/rabbitmq.ex

```diff
@@ -0,0 +1,124 @@
+defmodule Medusa.Adapter.RabbitMQ do
+  @moduledoc false
+  @behaviour Medusa.Adapter
+  use GenServer
+  use AMQP
+  require Logger
+  alias Medusa.Broker
+
+  defstruct channel: nil, routes: MapSet.new
+
+  defmodule Message do
+    defstruct event: nil, message: nil
+  end
+
+  @exchange_name "medusa"
+
+  def start_link do
+    GenServer.start_link(__MODULE__, [], name: __MODULE__)
+  end
+
+  def new_route(event) do
+    GenServer.call(__MODULE__, {:new_route, event})
+  end
+
+  def publish(event, message) do
+    GenServer.call(__MODULE__, {:publish, event, message})
+  end
+
+  def init([]) do
+    connection = init_connection()
+    {:ok, connection}
+  end
+
+  def handle_call({:new_route, event}, _from, state) do
+    Logger.debug "#{inspect __MODULE__}: [#{inspect event}]"
+    new_state = Map.update(state,
+                           :routes,
+                           MapSet.new([event]),
+                           &(MapSet.put(&1, event)))
+    {:reply, :ok, new_state}
+  end
+
+  def handle_call({:publish, event, payload}, _from, state) do
+    Logger.debug "#{inspect __MODULE__}: [#{inspect event}]: #{inspect payload}"
+    message = %Message{event: event, message: payload} |> Poison.encode!
+    AMQP.Basic.publish(state.channel, @exchange_name, "", message)
+    {:noreply, state}
+  end
+
+  def handle_info({:basic_consume_ok, %{consumer_tag: _consumer_tag}}, state) do
+    {:noreply, state}
+  end
+
+  def handle_info({:basic_cancel, %{consumer_tag: _consumer_tag}}, state) do
+    {:stop, :normal, state}
+  end
+
+  def handle_info({:basic_cancel_ok, %{consumer_tag: _consumer_tag}}, state) do
+    {:noreply, state}
+  end
+
+  def handle_info({:basic_deliver, payload, %{delivery_tag: tag}}, state) do
+    msg = Poison.decode!(payload)
+    event = Map.fetch!(msg, "event")
+    message = Map.fetch!(msg, "message")
+    message = %Broker.Message{body: message["body"], metadata: message["metadata"]}
+    Enum.each(state.routes, &Broker.maybe_route({&1, event, message}))
```

Comment:
IMO we should let rabbit to deal with routing and delivery as mentioned in comment about topic exchange